src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java

```diff
@@ -118,6 +116,96 @@ public synchronized void close() throws IOException {
             return;
         }
 
+        try {
+            stopGracefully();
+        } finally {
+            stopForcefully();
+        }
+    }
+
+    @SuppressFBWarnings("SWL_SLEEP_WITH_LOCK_HELD")
+    private void stopGracefully() {
+        if (!process.isAlive()) {
+            return;
+        }
+        int pid = ((ProcessImplForWin32) process).getPid();
+        boolean sentConsoleCtrlEvent = false;
+        synchronized (Kernel32.INSTANCE) {
+            // First, start a separate process that holds the console alive
+            // so that later gg can re-attach to this same console
+            Process holderProc;
+            try {
+                // Waits indefinitely for a keystroke
+                holderProc = new ProcessBuilder().command("cmd", "/C", "pause").start();
+            } catch (IOException e) {
+                logger.atError(STOP_GRACEFULLY_EVENT).cause(e)
+                        .log("Failed to start holder process. Cannot stop gracefully");
+                return;
+            }
+
+            Kernel32 k32 = Kernel32.INSTANCE;
+            try {
+                // Must detach from current console before attaching to another
+                if (!k32.FreeConsole()) {
+                    logger.atError(STOP_GRACEFULLY_EVENT).log("FreeConsole error {}", k32.GetLastError());
+                }
+                // calling attachConsole right after a process is launched will fail with invalid handle error
+                // waiting a bit ensures that we can attach to a process that just got launched.
+                try {
+                    Thread.sleep(500);
+                } catch (InterruptedException ignored) {
+                }
```

Comment:
can this come outside the sync block?

Reply:
moved this wait to process creation logic